waterbox/opera/bizhawk.cpp

```diff
@@ -0,0 +1,402 @@
+#include "bizhawk.hpp"
+#include <stdlib.h>
+#include <string>
+#include <libretro.h>
+#include <lr_input.h>
+#include <opera_vdlp.h>
+#include <opera_mem.h>
+#include <opera_cdrom.h>
+#include <opera_xbus.h>
+
+struct MemoryAreas 
+{
+	uint8_t* wram;
+  uint8_t* vram;
+};
+
+struct MemorySizes
+{
+	size_t wram;
+  size_t vram;
+};
+
+#define _MAX_AUDIO_SAMPLE_COUNT 4096
+
+std::string _biosFilePath;
+std::string _gameFilePath;
+std::string _fontFilePath;
+int _port1Type;
+int _port2Type;
+controllerData_t _port1Value;
+controllerData_t _port2Value;
+MemoryAreas _memoryAreas;
+MemorySizes _memorySizes;
+uint32_t* _videoBuffer;
+size_t _videoHeight;
+size_t _videoWidth;
+size_t _videoPitch;
+int _region;
+bool _isLagFrame;
+int _nvramChanged;
+
+#define _MAX_SAMPLES 4096
+#define _CHANNEL_COUNT 2
+int16_t _audioBuffer[_MAX_SAMPLES * _CHANNEL_COUNT];
+size_t _audioSamples;
+
+extern "C"
+{  
+	   void* xbus_cdrom_plugin(int   proc_, 	void* data_);
+	   void opera_cdrom_set_callbacks(opera_cdrom_get_size_cb_t get_size_,	opera_cdrom_set_sector_cb_t set_sector_,	opera_cdrom_read_sector_cb_t read_sector_);
+	   void opera_nvram_init(void *buf, const int bufsize);
+    void opera_lr_callbacks_set_audio_sample(retro_audio_sample_t cb);
+    void opera_lr_callbacks_set_audio_sample_batch(retro_audio_sample_batch_t cb);
+    void opera_lr_callbacks_set_environment(retro_environment_t cb);
+    void opera_lr_callbacks_set_input_poll(retro_input_poll_t cb);
+    void opera_lr_callbacks_set_input_state(retro_input_state_t cb);
+    void opera_lr_callbacks_set_log_printf(retro_log_printf_t cb);
+    void opera_lr_callbacks_set_video_refresh(retro_video_refresh_t cb);
+    RETRO_API void *retro_get_memory_data(unsigned id);
+    RETRO_API size_t retro_get_memory_size(unsigned id);
+				void retro_set_controller_port_device(unsigned port_, unsigned device_);
+				void retro_get_system_av_info(struct retro_system_av_info *info_);
+}
+
+void RETRO_CALLCONV retro_video_refresh_callback(const void *data, unsigned width, unsigned height, size_t pitch)
+{
+		// printf("Video %p, w: %u, h: %u, p: %lu\n", data, width, height, pitch);
+		_videoBuffer = (uint32_t*)data;
+		_videoWidth = width;
+		_videoHeight = height;
+		_videoPitch = pitch;
+}
+
+void RETRO_CALLCONV retro_log_printf_callback(enum retro_log_level level, const char *format, ...)
+{
+		va_list ap;
+		va_start(ap, format);
+		printf(format, ap);
+		va_end(ap);
+}
+
+size_t RETRO_CALLCONV retro_audio_sample_batch_callback(const int16_t *data, size_t frames)
+{
+	 memcpy(_audioBuffer, data, sizeof(int16_t) * frames * _CHANNEL_COUNT);
+ 	_audioSamples = frames;
+	 return frames;
+}
+
+void RETRO_CALLCONV retro_input_poll_callback()
+{
+	_isLagFrame = false;
+		// printf("Libretro Input Poll Callback Called:\n");
+}
+
+int16_t processController(const int portType, controllerData_t& portValue, const unsigned device, const unsigned index, const unsigned id)

```

Comment:
It looks like the entire function would be better using a `switch` instead of dozens of `if`s. Also, I assume you have some copy-paste errors in here as things like `if (RETRO_DEVICE_ID_JOYPAD_Y)` make no sense.